lib/Dialect/TTNN/IR/TTNNOps.cpp

```diff
@@ -598,6 +598,66 @@ ::mlir::LogicalResult mlir::tt::ttnn::ToMemoryConfigOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// ToLayoutOp
+//===----------------------------------------------------------------------===//
+
+// ToLayoutOp canonicalization
+// ToLayoutOp can be canonicalized if the previous op is also a ToLayoutOp. The
+// previous op can be merged with the next ToLayoutOp op if the previous op has
+// only one use. df - data format, l - layout, ms - memory space, tml - tensor
+// memory layout
+//
+//                |
+//      -----------------------
+//      |     ToLayoutOp      |                     |
+//      | df1, l1, ms1, tml1  |          -----------------------
+//      -----------------------          |     ToLayoutOp      |
+//                |                 -->  | df2, l1, ms2, tml1  |
+//                |                      -----------------------
+//      -----------------------                     |
+//      |     ToLayoutOp      |
+//      |      df2, ms2       |
+//      -----------------------
+//                |
+//
+::mlir::LogicalResult
+mlir::tt::ttnn::ToLayoutOp::canonicalize(ToLayoutOp toLayoutOp,
+                                         PatternRewriter &rewriter) {
+  // Get the input operand and verify that the previous op is toLayoutOp
+  ToLayoutOp previousToLayoutOp =
+      toLayoutOp.getOperand(0).getDefiningOp<ToLayoutOp>();
+
+  if (!previousToLayoutOp) {
+    return mlir::failure();
+  }
+
+  // Check if the parent op has only one use. We can only merge if the
+  // previous op has only single use.
+  if (!previousToLayoutOp->hasOneUse()) {
+    return mlir::failure();
+  }
+
+  // Replace the previous op with the merged ToLayoutOp
+  Value mergedToLayout = rewriter.replaceOpWithNewOp<ToLayoutOp>(
```

Comment:
😭 clang-tidy has such annoyances.  This is the error it's reporting:

```
/opt/ttmlir-toolchain/include/mlir/IR/OperationSupport.h:1013:5: error: Address of stack memory associated with temporary object of type '(lambda at /opt/ttmlir-toolchain/include/mlir/IR/OperationSupport.h:1003:27)' is still referred to by the stack variable 'state' upon returning to the caller.  This will be a dangling reference [clang-analyzer-core.StackAddressEscape,-warnings-as-errors]
 1013 |     return *properties.as<T *>();
      |     ^
/__w/tt-mlir/tt-mlir/lib/Dialect/TTNN/IR/TTNNOps.cpp:631:3: note: Taking false branch
  631 |   if (!previousToLayoutOp) {
      |   ^
/__w/tt-mlir/tt-mlir/lib/Dialect/TTNN/IR/TTNNOps.cpp:637:7: note: Assuming the condition is false
  637 |   if (!previousToLayoutOp->hasOneUse()) {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/__w/tt-mlir/tt-mlir/lib/Dialect/TTNN/IR/TTNNOps.cpp:637:3: note: Taking false branch
  637 |   if (!previousToLayoutOp->hasOneUse()) {
      |   ^
/__w/tt-mlir/tt-mlir/lib/Dialect/TTNN/IR/TTNNOps.cpp:645:7: note: '?' condition is false
  645 |       toLayoutOp.getDtypeAttr() ? toLayoutOp.getDtypeAttr()
      |       ^
/__w/tt-mlir/tt-mlir/lib/Dialect/TTNN/IR/TTNNOps.cpp:647:7: note: '?' condition is false
  647 |       toLayoutOp.getMemoryConfigAttr()
      |       ^
/__w/tt-mlir/tt-mlir/lib/Dialect/TTNN/IR/TTNNOps.cpp:642:26: note: Calling 'RewriterBase::replaceOpWithNewOp'
  642 |   Value mergedToLayout = rewriter.replaceOpWithNewOp<ToLayoutOp>(
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  643 |       previousToLayoutOp, toLayoutOp.getType(), previousToLayoutOp.getInput(),
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  644 |       toLayoutOp.getLayoutAttr(),
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~
  645 |       toLayoutOp.getDtypeAttr() ? toLayoutOp.getDtypeAttr()
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  646 |                                 : previousToLayoutOp.getDtypeAttr(),
      |                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  647 |       toLayoutOp.getMemoryConfigAttr()
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  648 |           ? toLayoutOp.getMemoryConfigAttr()
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  649 |           : previousToLayoutOp.getMemoryConfigAttr(),
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  650 |       toLayoutOp.getDevice());
      |       ~~~~~~~~~~~~~~~~~~~~~~~
```

The problem is that there is a bug in MLIR that ends up returning a stack address, obvs this is bad, but it's not our bug.  Clang-tidy is supposed to ignore headers (and bugs in them) that aren't ours, which is does... mostly, there is a bit of a grey area when it comes to this kind of reporting.  From what I gather after hours scouring clang-tidy related inter webs is that the bug isn't necessarily that a stack address was returned, but that a stack addressed is _used_ (potentially by our code).  It also might have something to do with the fact that this MLIR function is a template which ends up being instantiated (indirectly) by us.

Reply:
Anyway, I've seen errors like this before, perhaps this exact one, it's kinda hacky but the way we've worked around this in the past is to add a `// NOLINTNEXTLINE` just above the callsite here so that clang-tidy just doesn't traverse down this path.

Reply:
Yeah, //NOLINTNEXTLINE seems to skip lint check! Thanks!